ledger/complete/mtrie/trie/trie.go

```diff
@@ -391,69 +507,88 @@ func update(
 	payloads []ledger.Payload, // the payloads to be updated at the given paths
 	compactLeaf *node.Node, // a compact leaf node from its ancester, it could be nil
 	prune bool, // prune is a flag for whether pruning nodes with empty payload. not pruning is useful for generating proof, expecially non-inclusion proof
-) (n *node.Node, allocatedRegCountDelta int64, allocatedRegSizeDelta int64, lowestHeightTouched int) {
+	payloadStorage ledger.PayloadStorage, // for getting and setting the payloads stored on leaf nodes
+) (n *node.Node, allocatedRegCountDelta int64, allocatedRegSizeDelta int64, lowestHeightTouched int, err error) {
 	// No new path to update
 	if len(paths) == 0 {
 		if compactLeaf != nil {
+			leafHash := compactLeaf.ExpandedLeafHash()
+			compactPath, leafPayload, err := payloadStorage.Get(leafHash)
+			if err != nil {
+				return nil, 0, 0, 0, fmt.Errorf("could not get payload by leaf hash %v: %w", leafHash, err)
+			}
+
 			// if a compactLeaf from a higher height is still left,
 			// then expand the compact leaf node to the current height by creating a new compact leaf
 			// node with the same path and payload.
 			// The old node shouldn't be recycled as it is still used by the tree copy before the update.
-			n = node.NewLeaf(*compactLeaf.Path(), compactLeaf.Payload(), nodeHeight)
-			return n, 0, 0, nodeHeight
+			n = node.NewLeaf(compactPath, leafPayload, nodeHeight)
+			return n, 0, 0, nodeHeight, nil
 		}
 		// if no path to update and there is no compact leaf node on this path, we return
 		// the current node regardless it exists or not.
-		return currentNode, 0, 0, nodeHeight
+		return currentNode, 0, 0, nodeHeight, nil
 	}
 
-	if len(paths) == 1 && currentNode == nil && compactLeaf == nil {
+	if len(paths) == 1 && currentNode.IsDefaultNode() && compactLeaf == nil {
```

Comment:
This is a necessary change. 

In the case of `prune = false` while building a trie for generating proof, we might run into a case where currentNode is not `nil` but an empty node. If we only check `currentNode == nil`, then this if statement would be false, and end up trying to fetch an empty payload from storage and fail. 

So we need to check against `IsDefaultNode` to cover the case where currentNode is a compact leaf node with empty payload.